source/common/http/header_map_impl.cc

```diff
@@ -178,6 +179,78 @@ template <> bool HeaderMapImpl::HeaderList::isPseudoHeader(const LowerCaseString
   return key.get().c_str()[0] == ':';
 }
 
+namespace {
+
+// Loading the threshold for lazy map once to avoid changing it while an
+// header-map object is active, and to avoid costly calls to Runtime::getInteger.
+uint32_t lazyMapMinSize() {
+  CONSTRUCT_ON_FIRST_USE(
+      uint32_t, static_cast<uint32_t>(Runtime::getInteger("envoy.http.headermap.lazy_map_min_size",
+                                                          std::numeric_limits<uint32_t>::max())));
+}
+
+} // namespace
+
+bool HeaderMapImpl::HeaderList::maybeMakeMap() {
+  if (lazy_map_.empty()) {
+    if (headers_.size() < lazyMapMinSize()) {
+      return false;
+    }
+    // Add all entries from the list into the map.
+    for (auto node = headers_.begin(); node != headers_.end(); ++node) {
+#ifdef HEADERMAP_TYPE_MULTIMAP
```

Comment:
I would rather not have the ifdefs. Can we pick one and go with it? We can always change later? What's the current thinking on which to go with? Or do you want to make the type runtime configurable also so it can be experimented with at runtime?

Reply:
From the tests I've got so far, it seems that flat_hash_map, with a threshold of 3 is a good option.
I'd like to finalize the results once we agree on the instantiation mechanism, double check them, and then remove the ifdefs.

Reply:
sgtm

Reply:
Yes this SGTM. Constructor is fine, so let's run the numbers again and pick the one to use?

/wait

Reply:
Thanks! I'll run the tests, update the results and push the modified PR.